src/emqttd_session.erl

```diff
@@ -409,55 +410,59 @@ handle_cast({unsubscribe, _From, TopicTable},
                         SubMap
                 end
         end, Subscriptions, TopicTable),
-    noreply(emit_stats(State#state{subscriptions = Subscriptions1}));
+    hibernate(emit_stats(State#state{subscriptions = Subscriptions1}));
 
 %% PUBACK:
 handle_cast({puback, PacketId}, State = #state{inflight = Inflight}) ->
-    case Inflight:contain(PacketId) of
-        true ->
-            noreply(dequeue(acked(puback, PacketId, State)));
-        false ->
-            ?LOG(warning, "The PUBACK ~p is not inflight: ~p",
-                 [PacketId, Inflight:window()], State),
-            emqttd_metrics:inc('packets/puback/missed'),
-            noreply(State)
-    end;
+    {noreply,
+     case Inflight:contain(PacketId) of
+         true ->
+             dequeue(acked(puback, PacketId, State));
+         false ->
+             ?LOG(warning, "PUBACK ~p missed inflight: ~p",
+                  [PacketId, Inflight:window()], State),
+             emqttd_metrics:inc('packets/puback/missed'),
+             State
+     end, hibernate};
 
 %% PUBREC:
 handle_cast({pubrec, PacketId}, State = #state{inflight = Inflight}) ->
-    case Inflight:contain(PacketId) of
-        true ->
-            noreply(acked(pubrec, PacketId, State));
-        false ->
-            ?LOG(warning, "The PUBREC ~p is not inflight: ~p",
-                 [PacketId, Inflight:window()], State),
-            emqttd_metrics:inc('packets/pubrec/missed'),
-            noreply(State)
-    end;
+    {noreply,
+     case Inflight:contain(PacketId) of
+         true ->
+             acked(pubrec, PacketId, State);
+         false ->
+             ?LOG(warning, "PUBREC ~p missed inflight: ~p",
+                  [PacketId, Inflight:window()], State),
+             emqttd_metrics:inc('packets/pubrec/missed'),
+             State
+     end, hibernate};
 
 %% PUBREL:
 handle_cast({pubrel, PacketId}, State = #state{awaiting_rel = AwaitingRel}) ->
-    case maps:take(PacketId, AwaitingRel) of
-        {Msg, AwaitingRel1} ->
-            spawn(emqttd_server, publish, [Msg]),%%:)
-            noreply(State#state{awaiting_rel = AwaitingRel1});
-        error ->
-            ?LOG(warning, "Cannot find PUBREL: ~p", [PacketId], State),
-            emqttd_metrics:inc('packets/pubrel/missed'),
-            noreply(State)
-    end;
+    {noreply,
+     case maps:take(PacketId, AwaitingRel) of
+         {Msg, AwaitingRel1} ->
+             spawn(emqttd_server, publish, [Msg]), %%:)
```

Comment:
Measure the memory size emqttd_server:publish() may use. For example:
1) Set the max_heap_size to 1000, and kill = false, error_logger = true.
2) Run some cases and check the log that any max heap error is present. If present, it means max_heap_size is not big enough to held variables in these functions.
3) Increase max_heap_size to 2000.
4) Run same cases again, If no error, 2000 is a safe value.
5) min_heap_size should be at least 2000 within the spawn_opt.